escodegen.js

```diff
@@ -1836,8 +1838,34 @@
         BinaryExpression: function (expr, precedence, flags) {
             var result, leftPrecedence, rightPrecedence, currentPrecedence, fragment, leftSource;
             currentPrecedence = BinaryPrecedence[expr.operator];
-            leftPrecedence = expr.operator === '**' ? Precedence.Postfix : currentPrecedence;
-            rightPrecedence = expr.operator === '**' ? currentPrecedence : currentPrecedence + 1;
+
+            leftPrecedence = currentPrecedence;
+            rightPrecedence = currentPrecedence + 1;
+
+            switch (expr.operator) {
+                case '**':
+                    leftPrecedence = Precedence.Postfix;
+                    rightPrecedence = currentPrecedence;
+                    break;
+
+                case '??':
+                    if (expr.left.operator === '||' || expr.left.operator === '&&') {
```

Comment:
You shouldn't access the `operator` property here without first checking that node whose property you are accessing is a `LogicalExpression`:

```suggestion
                    if (expr.left.type === 'LogicalExpression' && (expr.left.operator === '||' || expr.left.operator === '&&')) {
```

Ditto below.

Reply:
thank you https://github.com/estools/escodegen/pull/417/commits/5646f113ae78eca9046da1b8e4ea20d4baec87ad

Reply:
Also added here:
https://github.com/estools/escodegen/pull/417/commits/7f97bfaaf618b515b45f1c2725a0452c8849c1d0